polaris-react/src/utilities/ephemeral-presence-manager/context.tsx

```diff
@@ -0,0 +1,18 @@
+import {createContext} from 'react';
+
+export type EphemeralPresenceKey = 'tooltip';
```

Comment:
Is there a specific reason why we'd want to lock down possible keys? Feels like this would make the manager harder to use outside of this intended effect. 

Relaxing the condition to simply take strings would immediately increase the possibility of key conflicts (several users of the manager clashing over the same string), which in some cases would be hard to resolve.

I was wondering if we could switch to using `Map` instead of objects thus being able to take anything as a key.

```
export interface EphemeralPresenceManagerContextType {
  presenceList: Map<any, boolean>;
  presenceCounter: Map<any, number>;
  addPresence: (key: any) => void;
  removePresence: (key: any) => void;
}

Then this could be used as:


Inside `Tooltip` itself
```
const {presenceList, addPresence, removePresence} = useEphemeralPresenceManager();

setShouldAnimate(!presenceList.has(this));
addPresence(this);
```

Outside of `Tooltip`
```
import {Tooltip} from 'polaris';
import {Tooltip as WebTooltip} from 'web/another-tooltip';

const {presenceList, addPresence, removePresence} = useEphemeralPresenceManager();

const hasPolarisTooltip = presenceList.has(Tooltip);
const hasWebTooltip = presenceList.has(WebTooltip);
```

I'm not 100% sure keys like this would work, but feels like some sort of API that uses the class we want to check presence of instead of a locked in convention might give us some additional flexibility.

Reply:
> [...] as this would become very tricky when working with functional components

Yeah, I was thinking we'd very much want to use `Symbol`s for this which could alleviate that problem a bit.

> I see this manager being concerned only with Polaris components, where we can control the keys easily, rather than allowing the consumer of this library to add additional presence keys

I guess that's fair enough! I wonder if we will need to extend or copy this presenceManager in admin somehow to cover the same use case but for components that are exclusively built there, but we can cross that bridge when we come to it!

> One thing we could do to help with that is to ensure that we export a different hook for the consumer, which had read-only access to the presenceList and presenceCounter but not the add/remove methods.

I'm not sure I see much difference... could you illustrate that better so I can get a better idea?

Reply:
Ah, I see what you mean now, presence is directly controlled within Polaris, so there should be no need to expose those to the outside.

One of the cases I had in mind for admin was [`GlobalToast`](https://github.com/Shopify/web/blob/main/app/foundation/Frame/components/GlobalToast/GlobalToast.tsx). I assume this should work since `Toast` would be the one informing of presence and not this global manager in admin, correct? 🤔

Reply:
Yeah, the Toast within GlobalToast would already handle the presence in that case.